ddprof-lib/src/main/cpp/os_linux_dd.cpp

```diff
@@ -42,4 +89,221 @@ SigAction ddprof::OS::replaceSigbusHandler(SigAction action) {
+  return signal_num;
+}
+
+void ddprof::OS::uninstallTlsPrimeSignalHandler(int signal_num) {
+  if (signal_num <= 0) {
+    return;
+  }
+  
+  struct sigaction sa;
+  sa.sa_handler = SIG_DFL;
+  sigemptyset(&sa.sa_mask);
+  sa.sa_flags = 0;
+  
+  if (sigaction(signal_num, &sa, NULL) != 0) {
+    TEST_LOG("Failed to uninstall RT signal %d for TLS priming: %s", 
+             signal_num, strerror(errno));
+  } else {
+    TEST_LOG("Successfully uninstalled TLS priming handler for RT signal %d", signal_num);
+  }
+}
+
+void ddprof::OS::enumerateThreadIds(const std::function<void(int)>& callback) {
+  DIR *task_dir = opendir("/proc/self/task");
+  if (!task_dir) {
+    TEST_LOG("Failed to open /proc/self/task: %s", strerror(errno));
+    return;
+  }
+  
+  struct dirent *entry;
+  while ((entry = readdir(task_dir)) != nullptr) {
+    if (entry->d_name[0] >= '1' && entry->d_name[0] <= '9') {
+      int tid = atoi(entry->d_name);
+      if (tid > 0) {
+        callback(tid);
+      }
+    }
+  }
+  closedir(task_dir);
+}
+
+void ddprof::OS::signalThread(int tid, int signum) {
+  if (syscall(SYS_tgkill, getpid(), tid, signum) != 0 && errno != ESRCH) {
+    TEST_LOG("Failed to signal thread %d with signal %d: %s", tid, signum, strerror(errno));
+  }
+}
+
+int ddprof::OS::getThreadCount() {
+  FILE *status = fopen("/proc/self/status", "r");
+  if (!status) {
+    return -1;
+  }
+  
+  char line[256];
+  int thread_count = -1;
+  while (fgets(line, sizeof(line), status)) {
+    if (sscanf(line, "Threads:\t%d", &thread_count) == 1) {
+      break;
+    }
+  }
+  fclose(status);
+  return thread_count;
+}
+
+bool ddprof::OS::startThreadDirectoryWatcher(const std::function<void(int)>& on_new_thread, const std::function<void(int)>& on_dead_thread) {
+  // Ensure fork handler is registered to prevent deadlock in child processes
+  ensureForkHandlerRegistered();
+
+  if (g_watcher_running.load()) {
+    return true; // Already running
+  }
+
+  int inotify_fd = inotify_init1(IN_CLOEXEC | IN_NONBLOCK);
+  if (inotify_fd == -1) {
+    TEST_LOG("Failed to initialize inotify: %s", strerror(errno));
+    return false;
+  }
+  
+  int watch_fd = inotify_add_watch(inotify_fd, "/proc/self/task", IN_CREATE | IN_DELETE | IN_MOVED_FROM | IN_MOVED_TO);
+  if (watch_fd == -1) {
+    TEST_LOG("Failed to add inotify watch on /proc/self/task: %s", strerror(errno));
+    close(inotify_fd);
+    return false;
+  }
+  
+  g_on_new_thread = on_new_thread;
+  g_on_dead_thread = on_dead_thread;
+  g_watcher_fd.store(inotify_fd);
+  g_watcher_running.store(true);
+  
+  if (pthread_create(&g_watcher_thread, nullptr, threadDirectoryWatcherLoop, nullptr) != 0) {
+    TEST_LOG("Failed to create thread directory watcher thread: %s", strerror(errno));
+    g_watcher_running.store(false);
+    g_watcher_fd.store(-1);
+    close(inotify_fd);
+    return false;
+  }
+
+  g_watcher_thread_created.store(true);
```

Comment:
Is this sufficient to avoid races? In other words, is this method only ever called non-concurrently? Because otherwise we have a window between pthread_create() and store() where a concurrent load() would still see false.